application/src/main/java/eu/europa/ec/fisheries/uvms/activity/service/facatch/evolution/CatchEvolutionProgressHandler.java

```diff
@@ -86,32 +87,33 @@ protected void handleCumulatedCatchWithDeletion(FaCatchEntity faCatch, CatchEvol
 
     protected void handleCumulatedCatch(FaCatchEntity faCatch, CatchSummaryListDTO cumulated, Map<String, Double> speciesCumulatedWeight, boolean delete) {
         String speciesCode = faCatch.getSpeciesCode();
-        boolean exists = false;
         cumulated.setTotal(delete ? cumulated.getTotal() - faCatch.getCalculatedWeightMeasure() : cumulated.getTotal() + faCatch.getCalculatedWeightMeasure());
 
-        for (SpeciesQuantityDTO speciesQuantityDTO : cumulated.getSpeciesList()) {
-            if (speciesQuantityDTO.getSpeciesCode().equalsIgnoreCase(speciesCode)) {
-                double weight = delete ? speciesQuantityDTO.getWeight() - faCatch.getCalculatedWeightMeasure() : speciesQuantityDTO.getWeight() + faCatch.getCalculatedWeightMeasure();
-                speciesQuantityDTO.setWeight(weight);
-
-                if (speciesCumulatedWeight.containsKey(speciesCode)) {
-                    double existingCumulatedWeight = speciesCumulatedWeight.get(speciesCode);
-                    double newCumulatedWeight = delete ? existingCumulatedWeight - faCatch.getCalculatedWeightMeasure() : existingCumulatedWeight + faCatch.getCalculatedWeightMeasure();
-                    speciesCumulatedWeight.put(speciesCode, newCumulatedWeight);
-                } else {
-                    speciesCumulatedWeight.put(speciesCode, weight);
-                }
-
-                exists = true;
-                break;
+        Optional<SpeciesQuantityDTO> previousAccumulationForThisSpeciesOptional = cumulated.getSpeciesList().stream()
+                .filter(dto -> speciesCode.equalsIgnoreCase(dto.getSpeciesCode()))
+                .findAny();
+
+        if (previousAccumulationForThisSpeciesOptional.isPresent()) {
+            SpeciesQuantityDTO previousAccumulationForThisSpecies = previousAccumulationForThisSpeciesOptional.get();
+            double weight = delete ? previousAccumulationForThisSpecies.getWeight() - faCatch.getCalculatedWeightMeasure() : previousAccumulationForThisSpecies.getWeight() + faCatch.getCalculatedWeightMeasure();
+            previousAccumulationForThisSpecies.setWeight(weight);
+
+            if (speciesCumulatedWeight.containsKey(speciesCode)) {
+                double existingCumulatedWeight = speciesCumulatedWeight.get(speciesCode);
+                double newCumulatedWeight = delete ? existingCumulatedWeight - faCatch.getCalculatedWeightMeasure() : existingCumulatedWeight + faCatch.getCalculatedWeightMeasure();
+                speciesCumulatedWeight.put(speciesCode, newCumulatedWeight);
+            } else {
+                speciesCumulatedWeight.put(speciesCode, weight);
             }
-        }
-
-        if (!exists) {
+        } else {
```

Comment:
Some tests would be nice before refactoring. Although it looks correct, more of a nice-to-have.

Reply:
It is at least covered by an Arquillian test